S21/jackzhenguo/refactorme.py

```diff
@@ -1,84 +1,112 @@
 import unicodedata
 
+
 class Word:
+    def __init__(self, word, bcp_code):
+        self.wc = word
+        self.lang = bcp_code
+
+
+class Strategy(object):
+    def __init__(self, word: Word):
+        self.word = word
+    """
+    strategy interface
+    """
+    def process(self):
+        pass
+
+
+class StrategyBase(Strategy):
+    def process(self) -> str:
+        """
+        strategy of only lower process
+        :return:
+        """
+        return self.word.wc.lower()
+
+
+class StrategyKeep(Strategy):
+    def process(self) -> str:
+        """
+        strategy of input word same to output word
+        :return:
+        """
+        return self.word.wc
+
+
+class StrategyVowel(Strategy):
+    """
+    strategy of processing to vowel cases
+    """
+    def process(self) -> str:
+        vowels = 'AEIOU\u00c1\u00c9\u00cd\u00d3\u00da'
+        temp = self.word.wc
+        if len(temp) > 1 and (self.word.wc[0] in ['t', 'n']) and \
+                unicodedata.normalize('NFC', self.word.wc)[1] in vowels:
+            temp = self.word.wc[0] + '-' + temp[1:]
+        return temp.lower()
+
+
+class StrategyCharReplace(Strategy):
+    def process(self) -> str:
+        """
+        strategy of replacing single char
+        :return:
+        """
+        temp = self.word.wc
+        temp = temp.replace('\u0049', '\u0131')
+        return temp.lower()
+
+
+class StrategyTrail(Strategy):
+    """
+    strategy of processing trail char
+    """
+    def process(self):
+        temp = self.word.wc
+        if temp[-1] == '\u03a3':
+            temp = temp[:-1] + '\u03c2'
+        return temp.lower()
+
+
+class WordProcess(object):
+    def __init__(self, word: Word):
+        self.word = word
+        self.strategy_dict = {'en': StrategyBase, 'zh': StrategyKeep, 'ja': StrategyKeep,
+                              'th': StrategyKeep, 'ga': StrategyVowel, 'tr': StrategyCharReplace,
+                              'az': StrategyCharReplace, 'el': StrategyTrail}
+
+    def to_lower(self):
+        language = self.convert_lang()
+        if language in self.strategy_dict:
+            specific_strategy = self.strategy_dict[language]
+        else:
+            specific_strategy = StrategyBase  # default strategy
+        return specific_strategy(self.word).process()
+
+    def convert_lang(self) -> str:
```

Comment:
I'd like this moved to it's own class representing BCP-47 strings.

Reply:
Thank you for giving me such good suggestions. Actually, we have done another version yesterday from our team that Willie pulled request.